File: frontend/src/components/benchmark/content/adamic_adar/implementations/adamic_adar.go
package main

import (
	"math"
	"sort"
)

type AdamicAdar struct {
	adjacencyList map[int]map[int]bool
}

func NewAdamicAdar() *AdamicAdar {
	return &AdamicAdar{
		adjacencyList: make(map[int]map[int]bool),
	}
}

func (aa *AdamicAdar) AddEdge(u, v int) {
	if aa.adjacencyList[u] == nil {
		aa.adjacencyList[u] = make(map[int]bool)
	}
	if aa.adjacencyList[v] == nil {
		aa.adjacencyList[v] = make(map[int]bool)
	}
	aa.adjacencyList[u][v] = true
	aa.adjacencyList[v][u] = true
}

func (aa *AdamicAdar) BuildFromEdges(edges [][2]int) {
	for _, edge := range edges {
		aa.AddEdge(edge[0], edge[1])
	}
}

func (aa *AdamicAdar) AdamicAdarScore(u, v int) float64 {
	if aa.adjacencyList[u] == nil || aa.adjacencyList[v] == nil {
		return 0.0
	}

	score := 0.0
	for node := range aa.adjacencyList[u] {
		if aa.adjacencyList[v][node] {
			degree := len(aa.adjacencyList[node])
			if degree > 1 {
				score += 1.0 / math.Log(float64(degree))
			}
		}
	}
	return score
}

func (aa *AdamicAdar) ScoreAllPairs() [][3]interface{} {
	var scores [][3]interface{}
	nodes := make([]int, 0, len(aa.adjacencyList))

	for node := range aa.adjacencyList {
		nodes = append(nodes, node)
	}

	for i := 0; i < len(nodes); i++ {
		for j := i + 1; j < len(nodes); j++ {
			u := nodes[i]
			v := nodes[j]
			score := aa.AdamicAdarScore(u, v)
			if score > 0 {
				scores = append(scores, [3]interface{}{u, v, score})
			}
		}
	}

	sort.Slice(scores, func(i, j int) bool {
		return scores[i][2].(float64) > scores[j][2].(float64)
	})

	return scores
}

func (aa *AdamicAdar) TopKPredictions(k int) [][3]interface{} {
	scores := aa.ScoreAllPairs()
	if k > len(scores) {
		k = len(scores)
	}
	return scores[:k]
}

File: frontend/src/components/benchmark/content/adamic_adar/implementations/adamic_adar.java
import java.util.*;

public class AdamicAdar {
    private Map<Integer, Set<Integer>> adjacencyList;

    public AdamicAdar() {
        this.adjacencyList = new HashMap<>();
    }

    public void addEdge(int u, int v) {
        adjacencyList.putIfAbsent(u, new HashSet<>());
        adjacencyList.putIfAbsent(v, new HashSet<>());
        adjacencyList.get(u).add(v);
        adjacencyList.get(v).add(u);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    public double adamicAdarScore(int u, int v) {
        if (!adjacencyList.containsKey(u) || !adjacencyList.containsKey(v)) {
            return 0.0;
        }

        double score = 0.0;
        for (int node : adjacencyList.get(u)) {
            if (adjacencyList.get(v).contains(node)) {
                int degree = adjacencyList.get(node).size();
                if (degree > 1) {
                    score += 1.0 / Math.log(degree);
                }
            }
        }
        return score;
    }

    public List<Map.Entry<int[], Double>> scoreAllPairs() {
        List<Map.Entry<int[], Double>> scores = new ArrayList<>();
        List<Integer> nodes = new ArrayList<>(adjacencyList.keySet());

        for (int i = 0; i < nodes.size(); i++) {
            for (int j = i + 1; j < nodes.size(); j++) {
                int u = nodes.get(i);
                int v = nodes.get(j);
                double score = adamicAdarScore(u, v);
                if (score > 0) {
                    scores.add(new AbstractMap.SimpleEntry<>(new int[]{u, v}, score));
                }
            }
        }

        scores.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        return scores;
    }

    public List<Map.Entry<int[], Double>> topKPredictions(int k) {
        List<Map.Entry<int[], Double>> scores = scoreAllPairs();
        return scores.subList(0, Math.min(k, scores.size()));
    }
}